### PR TITLE
[JIT] add support for multiple assignment statements

### PR DESCRIPTION
**Pull request by eellison**
add support for : `a = b, c = (1, 2)`

partial fix for https://github.com/pytorch/pytorch/issues/24256

### CODE REVIEW

File: test/test_jit.py [hunk 1]
```diff
@@ -10294,6 +10294,19 @@ def subscript_tuple_augmented_assign(a):
         with self.assertRaisesRegex(RuntimeError, 'does not support augmented assign'):
             scripted_aug_assign = torch.jit.script(subscript_tuple_augmented_assign)
 
+    def test_multiple_assign(self):
+        def test():
+            a = b, c = d, f = (1, 1)
```
Comment (zdevito) [inline test/test_jit.py]: I do not see any tests for lvalues that are not variables. Do they work? That was the primary reason for omitting this before.

Comment (eellison) [reply test/test_jit.py]: there is also pattern matching assignment in the test. what else were you thinking ?

**Comment by eellison** [comment]
any thoughts people

File: torch/csrc/jit/script/compiler.cpp [hunk 2]
```diff
@@ -2003,6 +2003,31 @@ struct to_ir {
...
+    if (stmt.lhs_list().size() == 1) {
+      return emitSingleAssignment(stmt);
+    }
+    // multiple assign & annotated type not supported in python
+    TORCH_INTERNAL_ASSERT(stmt.lhs_list().size() > 1 && !stmt.type().present());
+    // a = b = expr()
+    // the semantics of multiple assignment is that expr() is emitted once, then
+    // from left to right the assignments are made
+    const auto tmp_name =
+        std::string("$tmp_assign_") + std::to_string(tmp_count++);
```
Comment (driazati) [inline torch/csrc/jit/script/compiler.cpp]: We're starting to have a number of places that generate fake names (e.g. for list comprehensions), not necessary for this PR but we should think about consolidating those so we don't end up with collisions

https://github.com/pytorch/pytorch/blob/10c456417ce49cb5bfad8aba2bd7e3a23a83aef2/torch/csrc/jit/script/compiler.cpp#L1043

File: torch/csrc/jit/script/parser.cpp [hunk 3]
```diff
@@ -450,11 +450,24 @@ struct ParserImpl {
...
-      L.expect(TK_NEWLINE);
       if (maybeOp.value()->kind() == '=') {
+        std::vector<Expr> lhs_list = {lhs};
+        while (L.nextIf('=')) {
+          lhs_list.push_back(rhs);
+          rhs = parseExpOrExpTuple();
+        }
+        if (type.present()) {
+          throw ErrorReport(type.range())
+              << "Annotated multiple assignment is not well-defined in python";
```
Comment (driazati) [inline torch/csrc/jit/script/parser.cpp]: say `not supported` instead of `not well-defined`

**Review by driazati — APPROVED** [review]
Looks good, some nits

File: test/test_jit.py [hunk 4]
```diff
@@ -10268,6 +10268,19 @@ def subscript_tuple_augmented_assign(a):
...
+
+            # side effect
+            ten = torch.tensor(1)
+            ten1 = ten2 = ten.add_(1)
+
+            return a, b, c, d, f, ten, ten1, ten2
+
+        self.checkScript(test, ())
+        # check frontend
+        self.assertEqual(torch.jit.script(test)(), test())
```
[the remark this answers is not in the document]
Comment (eellison) [reply test/test_jit.py]: oop i thought that checkscript doesn't check python frontend, but that's wrong

File: torch/csrc/jit/script/parser.cpp (hunk 3 above)
Comment (eellison) [reply torch/csrc/jit/script/parser.cpp]: Sure, this is what i'm referencing fyi https://mypy.readthedocs.io/en/latest/type_inference_and_annotations.html

**Pull request closed without merging**